lib/versioning/VersioningRequestProcessor.js

```diff
@@ -0,0 +1,404 @@
+            return callback(null, [{ key, value }], versionId);
+        }
+        // need to get the master version to check if this is the master version
+        this.writeCache.get({ db, key }, logger, (err, data) => {
+            if (err && !err.ObjNotFound) {
+                return callback(err);
+            }
+            const versionId = request.options.versionId;
+            const versionKey = formatVersionKey(request.key, versionId);
+            const ops = [{ key: versionKey, value: request.value }];
+            if (Version.from(data).getVersionId() === versionId) {
+                // the version is also the master version: need to update both
+                ops.push({ key: request.key, value: request.value });
+            }
+            return callback(null, ops, versionId);
+        });
+        return null;
+    }
+
+
+    del(request, logger, callback) {
+        const { db, key, options } = request;
+        // no versioning or versioning configuration off
+        if (!(options && options.versionId)) {
+            return this.writeCache.batch({ db,
+                                           array: [{ key, type: 'del' }] },
+                                         logger, callback);
+        }
+        // version specific DELETE
+        return this.processVersionSpecificDelete(request, logger,
+            (err, array, vid) => {
+                if (err) {
+                    return callback(err);
+                }
+                return this.writeCache.batch({ db, array }, logger,
+                    err => callback(err, `{"versionId":"${vid}"}`));
+            });
+    }
+
+    /**
+     * Processes a version specific deleteObject request. This will create a
+     * batch of operations for deleting the specific version and marking the
+     * master version of the object as a place holder for deletion if the
+     * specific version is also the master version.
+     *
+     * @param {object} request - the request in original
+     *                           RepdConnection format { db, key
+     *                           [, value][, type], method, options }
+     * @param {object} logger - logger
+     * @param {function} callback - expect callback(err, batch, versionId)
+     * @return {any} - to finish the call
+     */
+    processVersionSpecificDelete(request, logger, callback) {
+        const { db, key, options } = request;
+        // deleting a specific version
+        this.writeCache.get({ db, key }, logger, (err, data) => {
+            if (err && !err.ObjNotFound) {
+                return callback(err);
+            }
+            // delete the specific version
+            const versionId = options.versionId;
+            const versionKey = formatVersionKey(key, versionId);
+            const ops = [{ key: versionKey, type: 'del' }];
+            // update the master version as PHD if it is the deleting version
+            if (Version.isPHD(data) ||
```

Comment:
I get the following trace at this point, when doing a batch delete of versioned objects:

```
/git/ironman/S3/node_modules/arsenal/lib/versioning/Version.js:46
        if (value.indexOf('isPHD') < 0) {
                 ^

TypeError: Cannot read property 'indexOf' of undefined
    at Function.isPHD (/git/ironman/S3/node_modules/arsenal/lib/versioning/Version.js:46:18)
    at writeCache.get (/git/ironman/S3/node_modules/arsenal/lib/versioning/VersioningRequestProcessor.js:390:25)
    at queue.(anonymous function).queue.forEach.callback (/git/ironman/S3/node_modules/arsenal/lib/versioning/WriteCache.js:117:28)
    at Array.forEach (native)
    at WriteCache._dequeue (/git/ironman/S3/node_modules/arsenal/lib/versioning/WriteCache.js:112:40)
    at wgm.get (/git/ironman/S3/node_modules/arsenal/lib/versioning/WriteCache.js:65:18)
    at subDb.get (/git/ironman/S3/node_modules/arsenal/lib/network/level-net/index.js:48:28)
    at /git/ironman/S3/node_modules/level-sublevel/shell.js:101:15
    at /git/ironman/S3/node_modules/level-sublevel/nut.js:121:19
```

The direct cause is that the WriteCache returns `ObjNotFound` and a `null` data (as expected with `ObjNotFound`) but we're skipping this error explicitly from error checking and we're using the `data` object afterwards.

I don't know what would be the correct fix here, @ttvi any clue?